### PR TITLE
Support omitting runtime version and code generation timestamp from generated files

### LINKED ISSUE #379: Generated files contain a timestamp that is more harmful than helpful
### Is your feature request related to a problem? Please describe.
Yes - the problem is that the file(s) generated include a comment showing the `Generation date` as a date/time. This makes it difficult to see what files have _really_ changed when re-running/updating the connected service, because _all_ files get changed at least to have an updated `Generation date`.

#### Describe the solution you'd like
I would like it to either:

- a) not write the `Generation date` comment at all
- b) have an option to disable writing the `Generation date` comment

#### Describe alternatives you've considered
None.

#### Additional context
Here are the top few lines of a generated file that show the `Generation date` comment:

``` csharp
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//     Runtime Version:4.0.30319.42000
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

// Generation date: 2/14/2024 5:49:22 PM
namespace MyServiceNamespace.Entities
{
```

### PR DESCRIPTION
**Pull request by gathogojr**
This pull request replaces https://github.com/OData/ODataConnectedService/pull/362 and fixes #379

It supports omitting runtime version and code generation timestamp from generated files. When the option is set to true, the auto-generated code disclaimer section will look as follows:
```csharp
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

```
as opposed to:
```csharp
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//     Runtime Version:4.0.30319.42000
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

// Generation date: 08.03.2021 19:55:12
```

The `no-timestamp` / `ti` option introduced in #362 is replaced with `omit-versioning-info` / `vi`.

No release of OData Connected Service or CLI has been done subsequent to #362 being merged so having a common flag to control whether runtime version and code generation timestamp should omitted was deemed to be a better option.

### Additional Details
This pull request also fixes an issue causing the builds to fail by upgrading the VS build tools:
image

### CODE REVIEW

**Comment by gathogojr** [comment]
@gregwinterstein @jphorv-bdo @rpallares FYI

File: src/Microsoft.OData.Cli/GenerateCommand.cs [hunk 1]
```diff
@@ -95,10 +95,10 @@ public GenerateCommand()
 
             this.AddOption(internalModifier);
 
-            Option noTimestamp = new Option<bool>(new[] { "--no-timestamp", "-nt" })
+            Option noTimestamp = new Option<bool>(new[] { "--omit-versioning-info", "-vi" })
```
Comment (habbes) [inline src/Microsoft.OData.Cli/GenerateCommand.cs:98]: Would `-ovi` be a better shorthand?

**Review by habbes — APPROVED** [review]

File: src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs [hunk 2]
```diff
@@ -397,9 +397,9 @@ public bool MakeTypesInternal
 }
 
 /// <summary>
-/// true to omit generation timestamp; otherwise false.
+/// true to omit runtime version and code generation timestamp; otherwise false.
```
Comment (ElizabethOkerio) [inline src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs:400]: nit: ```suggestion
/// Set to true to exclude the runtime version and code generation timestamp; otherwise, set to false.
```

File: src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs [hunk 3]
```diff
@@ -6312,24 +6322,28 @@ internal override void WriteFileHeader()
     {
 
 this.Write("\'------------------------------------------------------------------------------\r\n" +
-        "\' <auto-generated>\r\n\'     This code was generated by a tool.\r\n\'     Runtime Vers" +
-        "ion:");
+        "\' <auto-generated>\r\n\'     This code was generated by a tool.\r\n");
+
+
```
Comment (ElizabethOkerio) [inline src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs:6327]:
```suggestion
```

File: src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs [hunk 4]
```diff
@@ -6312,24 +6322,28 @@ internal override void WriteFileHeader()
...
-'     the code is regenerated.
-' </auto-generated>
-'------------------------------------------------------------------------------
-
-Option Strict Off
-Option Explicit On
-");
+this.Write("\r\n");
+
+
```
Comment (ElizabethOkerio) [inline src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs:6337]:
```suggestion
```

Comment (ElizabethOkerio) [reply src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs:6337]: I'm not sure about these spaces but I think they should be removed.

File: src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.tt [hunk 5]
```diff
@@ -36,8 +36,8 @@ public static class Configuration
 	// thereby making them invisible outside the assembly
 	public const bool MakeTypesInternal = false;
 
-	// If set to true, generation timestamps will be left out of generated files
-	public const bool NoTimestamp = false;
+	// If set to true, runtime version and code generation timestamp will be omitted from the generated files
```
Comment (ElizabethOkerio) [inline src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.tt:39]: In some instances the code string is different. Is it possible to have the same code string for this property throughout the code files.

File: src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs [hunk 6]
```diff
@@ -8629,7 +8643,7 @@ public class ODataT4CodeGeneratorBase
         /// <summary>
         /// The string builder that generation-time code is using to assemble generated output
         /// </summary>
-        protected System.Text.StringBuilder GenerationEnvironment
+        public System.Text.StringBuilder GenerationEnvironment
```
Comment (KenitoInc) [inline src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs:8646]: Why this change?

**Review by KenitoInc — COMMENTED** [review]
Just one question.
Otherwise LGTM

File: src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.tt (hunk 5 above)
Comment (gathogojr) [reply src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.tt]: @ElizabethOkerio I have addressed this in the following manner:
- For properties (with getter and setter), I followed the common convention for boolean properties and adopted the following text: `Gets or sets a value indicating whether to omit runtime version and code generation timestamp from the generated files.`
- For constants, I adopted the following text: `This flag indicates whether to omit runtime version and code generation timestamp from the generated files`
- For the UI label and CLI option, I adopted the following text: `Omit runtime version and code generation timestamp from the generated files.`

The texts reads the same tarting from the word "omit" but ahead of that I appended text that fits best in the particular usage

File: src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs (hunk 2 above)
Comment (gathogojr) [reply src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs]: @ElizabethOkerio See my comment here https://github.com/OData/ODataConnectedService/pull/383#discussion_r1519322063

File: src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs (hunk 4 above)
Comment (gathogojr) [reply src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs:6337]: @ElizabethOkerio This file is generated via the "Run Custom Tool" option in Visual Studio against the `ODataT4CodeGenerator.tt` file. Here's how the code that generates this output looks in the `ODataT4CodeGenerator.ttinclude` file:
```csharp
// <auto-generated>
//     This code was generated by a tool.
<#+
    if (!this.context.OmitVersioningInfo)
    {
#>
//     Runtime Version:<#= Environment.Version #>
<#+
    }
#>
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

<#+
    if (!this.context.OmitVersioningInfo)
    {
#>
// Generation date: <#= DateTime.Now.ToString(global::System.Globalization.CultureInfo.CurrentCulture) #>
<#+
    }
```
If we're to choose between keeping the logic in the `ttinclude` file legible plus easy to follow on the one side, and extra empty lines in the generated file, I'd rather the former.

File: src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs (hunk 3 above)
Comment (gathogojr) [reply src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs:6327]: @ElizabethOkerio See my comment here https://github.com/OData/ODataConnectedService/pull/383#discussion_r1519337199

File: src/Microsoft.OData.Cli/GenerateCommand.cs (hunk 1 above)
Comment (gathogojr) [reply src/Microsoft.OData.Cli/GenerateCommand.cs:98]: @habbes @KenitoInc We seem to have no specific convention at in respect to crafting the shorthand:
- `file-name` = `fn` --- I think this should have been just `f`. filename is a word...
- `outputdir` = `o` --- I find this okay. The shorter the better
- `metadata-uri` = `m` --- I find this okay too.
- `custom-headers` = `h` --- I find this okay too.
- `namespace-prefix` = `ns` --- Not `np` or `nsp`.  --- I find `ns` okay. Would have been fine with `n`.
- `upper-camel-case` = `ucc` --- IMO, a shorter form like `u` would have sufficed - so long as it's not being used as a short-hand for another option.
- `enable-internal` = `i` --- I find this okay too.
- `multiple-files` has no short-hand!!!
- `excluded-bound-operations` = `ebo` --- IMO, a shorter form like `eo` would have sufficed (i.e., `e` for exclude, while the second letter disambiguate between bound operations, operation imports and schema types).
- `excluded-operation-imports` = `eoi` --- IMO, a shorter form like `ei` would have sufficed.
- `excluded-schema-types` = `est` --- IMO, a shorter form like `et` would have sufficed.

The point I'm trying to illustrate with this is that we're not consistently combining the first letters from the option when crafting the shorthand. It's not a practice that I find common in any case. In my experience the shorthand of the option is seems to be typically kept as short as possible - one letter if possible, two letters to disambiguate.

**Review by ElizabethOkerio — APPROVED** [review]

**Review by KenitoInc — APPROVED** [review]

**Comment by jphorv-bdo** [comment]
Thank you everyone for your work on this, I look forward to seeing it published!

**Comment by jphorv-bdo** [comment]
> Thank you everyone for your work on this, I look forward to seeing it published!

Any estimate as to when this will be published to VS Marketplace?

**Comment by jphorv-bdo** [comment]
@ElizabethOkerio @KenitoInc @habbes  I just realized there have been 38 PRs merged since the last time a release was published to VS Marketplace on August 10, 2022.

I wish I had looked at that earlier - if I had, I wouldn't have bothered checking for a new release every day for the past month!

When will a new release be published to VS Marketplace?
